src/client/http.ts

```diff
@@ -101,11 +102,92 @@ export class HttpClient implements Requester {
   }
 
   public async request<TResult>(request: UpstashRequest): Promise<UpstashResponse<TResult>> {
+    const [url, requestOptions] = this.prepareRequest(request)
+
+    let response: Response | undefined = undefined;
+    let error: Error | undefined = undefined;
+    for (let index = 0; index < this.retry.attempts; index++) {
+      try {
+        response = await fetch(url.toString(), requestOptions);
+        break;
+      } catch (error_) {
+        error = error_ as Error;
+        await new Promise((r) => setTimeout(r, this.retry.backoff(index)));
+      }
+    }
+    if (!response) {
+      throw error ?? new Error("Exhausted all retries");
+    }
+    await this.checkResponse(response);
+
+    if (request.parseResponseAsJson === false) {
+      return undefined as unknown as UpstashResponse<TResult>;
+    }
+    return (await response.json()) as UpstashResponse<TResult>;
+  }
+
+  public async *requestStream(request: UpstashRequest): AsyncIterable<ChatCompletionChunk> {
+    const [url, requestOptions] = this.prepareRequest(request)
+
+    let response: Response | undefined = undefined;
+    let error: Error | undefined = undefined;
+    for (let index = 0; index < this.retry.attempts; index++) {
+      try {
+        response = await fetch(url, requestOptions);
+        break;
+      } catch (error_) {
+        error = error_ as Error;
+        await new Promise((r) => setTimeout(r, this.retry.backoff(index)));
+      }
+    }
+
+    if (!response) {
+      throw error ?? new Error("Exhausted all retries");
+    }
+    await this.checkResponse(response);
+
+    if (!response.body) {
+      throw new Error("No response body");
+    }
+
+    const reader = response.body.getReader();
+    const decoder = new TextDecoder();
+
+    try {
+      while (true) {
+        const { done, value } = await reader.read();
+        if (done) {
+          // stops here when max token reached
+          break;
+        }
+        
+        const chunkText = decoder.decode(value, { stream: true });
+        const chunks = chunkText.split('\n').filter(Boolean);
+
+        for (const chunk of chunks) {
+          if (chunk.startsWith('data: ')) {
+            const data = chunk.slice(6);
+
+            if (data === "[DONE]") {
+              // stops here last message is delivered
+              break;
+            }
+            
+            yield JSON.parse(data);
+          }
+        }
+      }
+    } finally {
+      await reader.cancel();
+    }
+  }
```

Comment:
Once the part above turned into a util function we can call this one like `processStream`, `parseStream` etc...

Reply:
named this method as processRequest. didn't use processStream because it serves both stream and non-stream requests.